shared-libs/rules-engine/src/contact-state-store.js

```diff
@@ -0,0 +1,209 @@
+/**
+ * @module contact-state-store
+ * In-memory datastore containing details on the state of each contact's rules calculations
+ */
+const md5 = require('md5');
+
+const EXPIRE_CALCULATION_AFTER_MS = 7 * 24 * 60 * 60 * 1000;
+let state, currentUser, onStateChange;
+
+const self = {
+  /**
+   * Initializes the contact-state-store from an existing state. If existing state is invalid, builds an empty state.
+   *
+   * @param {Object} existingState State object previously passed to the stateChangeCallback
+   * @param {Object} settingsDoc Settings document
+   * @param {Object} userDoc User's hydrated contact document
+   * @param {Object} stateChangeCallback Callback which is invoked whenever the state changes. Receives the updated state as the only parameter.
+   */
+  load: (existingState, settingsDoc, userDoc, stateChangeCallback) => {
+    if (state) {
+      throw Error('Attempted to initialize the contact-state-store multiple times.');
+    }
+
+    const rulesConfigHash = hashRulesConfig(settingsDoc, userDoc);
+    const useState = existingState && existingState.rulesConfigHash === rulesConfigHash;
+    if (!useState) {
+      return self.build(settingsDoc, userDoc, stateChangeCallback);
+    }
+
+    state = existingState;
+    currentUser = userDoc;
+    onStateChange = safeCallback(stateChangeCallback);
+  },
+
+  /**
+   * Initializes an empty contact-state-store.
+   *
+   * @param {Object} settingsDoc Settings document
+   * @param {Object} userDoc User's hydrated contact document
+   * @param {Object} stateChangeCallback Callback which is invoked whenever the state changes. Receives the updated state as the only parameter.
+   */
+  build: (settingsDoc, userDoc, stateChangeCallback) => {
+    if (state) {
+      throw Error('Attempted to initialize the contact-state-store multiple times.');
+    }
+
+    state = {
+      rulesConfigHash: hashRulesConfig(settingsDoc, userDoc),
+      contactStates: {},
+    };
+    currentUser = userDoc;
+
+    onStateChange = safeCallback(stateChangeCallback);
+    return onStateChange(state);
```

Comment:
It actually is not undefined. The callback may return a promise, as [our concrete implementation does](https://github.com/medic/cht-core/pull/6041/files#diff-73cb3dd3f8d28408c50f89304011bf43R199). Passing this promise down to the caller gives the option to have code wait for the state change to commit before it proceeds to the next task. For example, we do that [here](https://github.com/medic/cht-core/pull/6041/files#diff-73cb3dd3f8d28408c50f89304011bf43R87) but choose not to [here](https://github.com/medic/cht-core/pull/6041/files#diff-73cb3dd3f8d28408c50f89304011bf43R118). 

If the state fails to commit, downstream code in the promise will abort.
This also makes integration testing a lot cleaner, since we can easily await the state change to commit before checking the content of the document.

I personally feel this is something we should keep, though I agree it may be possible to do without.